### ISSUE TITLE
Allow to match a segment to one of several constraints

### ISSUE THREAD

**Issue by Domino987** [markphelps/flipt#180]
**Is your feature request related to a problem? Please describe.**
I need to toggle a feature in response to a certain parameter, e.g. I have 5 admin emails and if the API call sends one of them in the validate call, the feature should be toggled on by responding with true.
Currently I create a segment for each with the constraint email to be equal and the flag returns true for these segments. But setting several emails to the segment as constraints, since it checks if the given email is equal to all of them does not work.  Only if just one email is provided.

**Describe the solution you'd like**
Maybe set the option of the segment from always match all to be switchable to match one of. That way, I would only need to create one segment, give them all the emails as constraints and if the email matches one of the constraints, its in that segment and the feature flag would return true.

**Additional context**
I really enjoy your work and it is a great help. Thank you.

**Comment by markphelps** [comment]
@Domino987 thanks for the feature request. I think this would be doable, although it would require a bit of work because it would be changing Segments to work from 'Match All', to making them configurable via a radio button or something (in the UI) to be 'Match All' OR 'Match Any'

Here's what I'm thinking would require changes:

* [ ] UI as mentioned above when setting up a Segment
* [ ] DB migration to add new 'type' column for Segments (Match All or Match Any)
* [ ] API changes to allow for this new 'type' property of a Segment (default would be Match All I guess)
* [ ] Changes to evaluation logic to support 'Match Any'
* [ ] Unit and Integration Tests

I just want to make sure that this change doesn't slow down the evaluation logic, so I may need to add some benchmarks before and after as well.

I'll wip up a mock screen shot of what I think the Segment UI could look like and post it in this issue hopefully tonight.

Then I can go about estimating how long this would take to actually implement. 

**tl;dr:** This isn't a simple change, but I think it's doable and would be valuable to implement for use cases such as yours.

Please let me know if you have any additional thoughts or how you envisioned this to be exposed in the UI. Otherwise, watch this issue for updates!

Cheers 🍻

**Comment by Domino987** [comment]
@markphelps , wow you gave that already a lot of thought. thank you. I would personally use a drop down to use less space and make it possible to add more types if needed without changing the interface. 
I think it should not effect the evaluation  logic since it would be a switch to filter in dependance to the selected type for that segment and the match any should be faster with a short circuit once one match is found. 

I think that would be a great addition to flipt and make it more convenient for a lot of user.
Unfortunately, I can't help you much with the implementation since I mostly write js/ts and a python.

**Comment by markphelps** [comment]
@Domino987 I think im gonna go with a simple radio button like this, since really the only options for matching constraints are an AND or an OR (all vs any respectively)

![Kapture 2019-11-19 at 20 05 22](https://user-images.githubusercontent.com/209477/69200076-0df49c80-0b08-11ea-95fd-8b318859f0f0.gif)

**Comment by Domino987** [comment]
@markphelps , looks good. Will that automatically save the change to the db or will there be an additional save button. You could also move that to the segment creation dialog. But I like this approach as well, makes it easier to change.

**Comment by markphelps** [comment]
Good question.. I think I may actually move it above into the form so that it's clear that you need to hit the save button below in order to change the 'Match Type'. Something like this:

![Screen Shot 2019-11-21 at 9 04 58 AM](https://user-images.githubusercontent.com/209477/69344864-21098880-0c3e-11ea-9fc4-d127c01e752a.png)

What do you think?

**Comment by Domino987** [comment]
Looks good, That makes it consistent not one does not accidentally change the segment :)

**Comment by markphelps** [comment]
Tracking in https://github.com/markphelps/flipt/compare/segment-type

**Comment by markphelps** [comment]
@Domino987 this has been added in [v0.11.0](https://github.com/markphelps/flipt/releases/tag/v0.11.0)! Please give it a try and let me know what you think, and thank you again for the feature request!

**Comment by Domino987** [comment]
Hi looks good but I found one major error:
It will match the user to a segment, if the segment has no constraints. So maybe you should also check, if the segment has any constraints before matching it.
![image](https://user-images.githubusercontent.com/17567991/69951566-c4d41d80-14f5-11ea-8dbf-b978d6109506.png)


Every user will match to this segment now, which is not desired, but should be an easy fix.

Thank you !!
